gateway-service-metadata/src/main/java/org/apache/knox/gateway/service/metadata/ServiceModel.java

```diff
@@ -101,14 +103,27 @@ public String getContext() {
   }
 
   @XmlElement
-  public String getServiceUrl() {
+  public String getServiceUrl() throws MalformedURLException {
     String context = getContext();
     if ("HIVE".equals(getServiceName())) {
       return String.format(Locale.ROOT, "jdbc:hive2://%s:%d/;?hive.server2.transport.mode=http;hive.server2.thrift.http.path=/%s/%s%s", request.getServerName(),
           request.getServerPort(), gatewayPath, topologyName, context);
     } else {
+      final String backendUrlString = getBackendServiceUrl();
       if (context.indexOf("{{BACKEND_HOST}}") > -1) {
-        context = context.replace("{{BACKEND_HOST}}", getBackendServiceUrl());
+        context = context.replace("{{BACKEND_HOST}}", backendUrlString);
+      }
+      if (context.indexOf("{{SCHEME}}") > -1 || context.indexOf("{{HOST}}") > -1 || context.indexOf("{{PORT}}") > -1) {
+        final URL backendUrl = new URL(backendUrlString);
+        if (context.indexOf("{{SCHEME}}") > -1) {
+          context = context.replace("{{SCHEME}}", backendUrl.getProtocol());
+        }
+        if (context.indexOf("{{HOST}}") > -1) {
+          context = context.replace("{{HOST}}", backendUrl.getHost());
+        }
+        if (context.indexOf("{{PORT}}") > -1) {
+          context = context.replace("{{PORT}}", String.valueOf(backendUrl.getPort()));
+        }
```

Comment:
Not sure we need all the double checks? Can't we just do `context.replace(...)` and if it doesn't match it won't replace? The strings here are super short. It might even be slower to do a check and then replace anyway.

Reply:
We definetly need the first if (backend URL may start with unknown protocol, such as dfs, ws, etc..) -> result in an error.
I agree with the small ifs.

Reply:
Fixed; I'll wait till Travis finishes and merge this one.